autogpt/agent/LLM/openai_provider.py

```diff
@@ -0,0 +1,155 @@
+"""The base class for all LLM providers."""
+import enum
+import time
+
+import openai
+import pydantic
+from colorama import Fore, Style
+from openai.error import APIError, RateLimitError
+
+from autogpt.agent.LLM.base_llm_provider import BaseLLMProvider
+from autogpt.logs import logger
+
+
+class ChatCompletionModels(enum.Enum):
+    """Available Chat Completion Models"""
+
+    GTP_3_5_TURBO = "gpt-3.5-turbo"
+    GTP_3_5_TURBO_0301 = "gpt-3.5-turbo-0301"
+    GTP_4 = "gpt-4"
+    GTP_4_0314 = "gpt-4-0314"
+    GTP_4_32K = "gpt-4-32k"
+    GTP_4_32K_0314 = "gpt-4-32k-0314"
+
+
+class EmbeddingModels(enum.Enum):
+    """Available Embedding Models"""
+
+    TEXT_EMBEDDING_ADA_002 = "text-embedding-ada-002"
+    TEXT_SEARCH_ADA_DOC_001 = "text-search-ada-doc-001"
+
+
+class OpenAIProvider(BaseLLMProvider):
+    """The base class for all LLM providers."""
+
+    def __init__(
+        self,
+        api_key: str,
+        max_tokens: int = 1000,
+        chat_completion_model: ChatCompletionModels = ChatCompletionModels.GTP_4,
+        embedding_model: EmbeddingModels = EmbeddingModels.TEXT_EMBEDDING_ADA_002,
+        temperature: float = 0.9,
+        num_retries: int = 10,
+        debug_mode: bool = False,
+    ) -> None:
+        """Initialise the OpenAIProvider"""
+        self.api_key = api_key
+        self.max_tokens = max_tokens
+        self.chat_completion_model = chat_completion_model
+        self.embedding_model = embedding_model
+        self.temperature = temperature
+        self.debug_mode = debug_mode
+        self.num_retries = num_retries
+        self.warned_user = False
+        openai.api_key = api_key
+
+    def log_debug_msg(self, message: str) -> None:
+        """Log a debug message"""
+        if self.debug_mode:
+            logger.debug(Fore.GREEN + message + Fore.RESET)
+
+    async def attempt_chat_completion(self, messages: list, attempt: int = 0) -> str:
+        """Attempt to create a chat completion"""
+
+        try:
+            return await openai.ChatCompletion.acreate(
```

Comment:
Has this been tested? If `.acreate` is asyncio-flavoured, we need to convert it to trio-flavoured

Throw "call an asyncio-flavoured function within trio code" into GPT4.